File: alljoyn_core/src/winrt/binding/src/SessionOpts.cc
/******************************************************************************
 *
 * Copyright (c) 2011-2012, AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *****************************************************************************/

#include "SessionOpts.h"

#include <qcc/String.h>
#include <qcc/winrt/utility.h>
#include <alljoyn/Session.h>
#include <alljoyn/Status.h>
#include <ObjectReference.h>
#include <AllJoynException.h>

namespace AllJoyn {

SessionOpts::SessionOpts()
{
    ::QStatus status = ER_OK;

    while (true) {
        // Create managed _SessionOpts
        _mSessionOpts = new qcc::ManagedObj<_SessionOpts>();
        // Check for allocation error
        if (NULL == _mSessionOpts) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Store pointer to _SessionOpts for convenience
        _sessionOpts = &(**_mSessionOpts);
        break;
    }

    // Bubble up any QStatus errors as an exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

SessionOpts::SessionOpts(TrafficType traffic, bool isMultipoint, ProximityType proximity, TransportMaskType transports)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Create managed _SessionOpts
        ajn::SessionOpts::TrafficType trafficType = (ajn::SessionOpts::TrafficType)(int)traffic;
        ajn::SessionOpts::Proximity prox = (ajn::SessionOpts::Proximity)(int)proximity;
        ajn::TransportMask trans = (ajn::TransportMask)(int)transports;
        _mSessionOpts = new qcc::ManagedObj<_SessionOpts>(trafficType, isMultipoint, prox, trans);
        // Check for allocation error
        if (NULL == _mSessionOpts) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Store pointer to _SessionOpts for convenience
        _sessionOpts = &(**_mSessionOpts);
        break;
    }

    // Bubble up any QStatus errors as an exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

SessionOpts::SessionOpts(const ajn::SessionOpts* sessionOpts)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Check sessionOpts for invalid values
        if (NULL == sessionOpts) {
            status = ER_BAD_ARG_1;
            break;
        }
        // Create managed _SessionOpts
        _mSessionOpts = new qcc::ManagedObj<_SessionOpts>(sessionOpts->traffic,
                                                          sessionOpts->isMultipoint,
                                                          sessionOpts->proximity,
                                                          sessionOpts->transports);
        // Check for allocation error
        if (NULL == _mSessionOpts) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Store pointer to _SessionOpts for convenience
        _sessionOpts = &(**_mSessionOpts);
        break;
    }

    // Bubble up any QStatus errors as an exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

SessionOpts::SessionOpts(const qcc::ManagedObj<_SessionOpts>* sessionOpts)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Check sessionOpts for invalid values
        if (NULL == sessionOpts) {
            status = ER_BAD_ARG_1;
            break;
        }
        // Attach sessionOpts to managed _SessionOpts
        _mSessionOpts = new qcc::ManagedObj<_SessionOpts>(*sessionOpts);
        // Check for allocation error
        if (NULL == _mSessionOpts) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        // Store pointer to _SessionOpts for convenience
        _sessionOpts = &(**_mSessionOpts);
        break;
    }

    // Bubble up any QStatus errors as an exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

SessionOpts::~SessionOpts()
{
    // Delete managed _SessionOpts to adjust ref count
    if (NULL != _mSessionOpts) {
        delete _mSessionOpts;
        _mSessionOpts = NULL;
        _sessionOpts = NULL;
    }
}

TrafficType SessionOpts::Traffic::get()
{
    TrafficType result = (TrafficType)(int)-1;

    // Check if wrapped value already exists
    if ((TrafficType)(int)-1 == _sessionOpts->_eventsAndProperties->Traffic) {
        result = (TrafficType)(int)_sessionOpts->traffic;
        // Store the result
        _sessionOpts->_eventsAndProperties->Traffic = result;
    } else {
        // Return Traffic
        result = _sessionOpts->_eventsAndProperties->Traffic;
    }

    return result;
}

void SessionOpts::Traffic::set(TrafficType value)
{
    // Store value
    _sessionOpts->traffic = (ajn::SessionOpts::TrafficType)(int)value;
    _sessionOpts->_eventsAndProperties->Traffic = value;
}

bool SessionOpts::IsMultipoint::get()
{
    bool result = (bool)-1;

    // Check if wrapped value already exists
    if ((bool)-1 == _sessionOpts->_eventsAndProperties->IsMultipoint) {
        result = _sessionOpts->isMultipoint;
        // Store the result
        _sessionOpts->_eventsAndProperties->IsMultipoint = result;
    } else {
        // Return IsMultipoint
        result = _sessionOpts->_eventsAndProperties->IsMultipoint;
    }

    return result;
}

void SessionOpts::IsMultipoint::set(bool value)
{
    // Store value
    _sessionOpts->isMultipoint = value;
    _sessionOpts->_eventsAndProperties->IsMultipoint = value;
}

ProximityType SessionOpts::Proximity::get()
{
    ProximityType result = (ProximityType)(int)-1;

    // Check if wrapped value already exits
    if ((ProximityType)(int)-1 == _sessionOpts->_eventsAndProperties->Proximity) {
        result = (ProximityType)(int)_sessionOpts->proximity;
        // Store the result
        _sessionOpts->_eventsAndProperties->Proximity = result;
    } else {
        // Return Proximity
        result = _sessionOpts->_eventsAndProperties->Proximity;
    }

    return result;
}

void SessionOpts::Proximity::set(ProximityType value)
{
    // Store value
    _sessionOpts->proximity = (ajn::SessionOpts::Proximity)(int)value;
    _sessionOpts->_eventsAndProperties->Proximity = value;
}

TransportMaskType SessionOpts::TransportMask::get()
{
    TransportMaskType result = (TransportMaskType)(int)-1;

    // Check if wrapped value already exists
    if ((TransportMaskType)(int)-1 == _sessionOpts->_eventsAndProperties->TransportMask) {
        result = (TransportMaskType)(int)_sessionOpts->transports;
        // Store the result
        _sessionOpts->_eventsAndProperties->TransportMask = result;
    } else {
        // Return TransportMask
        result = _sessionOpts->_eventsAndProperties->TransportMask;
    }

    return result;
}

void SessionOpts::TransportMask::set(TransportMaskType value)
{
    // Store value
    _sessionOpts->transports = (ajn::TransportMask)(int)value;
    _sessionOpts->_eventsAndProperties->TransportMask = value;
}

_SessionOpts::_SessionOpts()
    : ajn::SessionOpts()
{
    ::QStatus status = ER_OK;

    while (true) {
        // Create internal ref class
        _eventsAndProperties = ref new __SessionOpts();
        // Check for allocation error
        if (nullptr == _eventsAndProperties) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        break;
    }

    // Bubble up any QStatus errors as an exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

_SessionOpts::_SessionOpts(ajn::SessionOpts::TrafficType traffic, bool isMultipoint, ajn::SessionOpts::Proximity proximity, ajn::TransportMask transports)
    : ajn::SessionOpts(traffic, isMultipoint, proximity, transports)
{
    ::QStatus status = ER_OK;

    while (true) {
        // Create internal ref class
        _eventsAndProperties = ref new __SessionOpts();
        // Check for allocation error
        if (nullptr == _eventsAndProperties) {
            status = ER_OUT_OF_MEMORY;
            break;
        }
        break;
    }

    // Bubble up any QStatus errors as an exception
    if (ER_OK != status) {
        QCC_THROW_EXCEPTION(status);
    }
}

_SessionOpts::~_SessionOpts()
{
    _eventsAndProperties = nullptr;
}

__SessionOpts::__SessionOpts()
{
    Traffic = (TrafficType)(int)-1;
    IsMultipoint = (bool)-1;
    Proximity = (ProximityType)(int)-1;
    TransportMask = (TransportMaskType)(int)-1;
}

__SessionOpts::~__SessionOpts()
{
    Traffic = (TrafficType)(int)-1;
    IsMultipoint = (bool)-1;
    Proximity = (ProximityType)(int)-1;
    TransportMask = (TransportMaskType)(int)-1;
}

}

File: alljoyn_java/src/org/alljoyn/bus/SessionOpts.java
/*
 * Copyright (c) 2009-2011, 2014 AllSeen Alliance. All rights reserved.
 *
 *    Permission to use, copy, modify, and/or distribute this software for any
 *    purpose with or without fee is hereby granted, provided that the above
 *    copyright notice and this permission notice appear in all copies.
 *
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 *    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 *    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 *    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 *    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 *    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 *    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

package org.alljoyn.bus;

public class SessionOpts {

    /**
     * Holds the traffic type for this SessionOpt
     */
    public byte traffic;

    /**
     * Multi-point session capable.
     *
     * A session is multi-point if it can be joined multiple times to form a single
     * session with multi (greater than 2) endpoints. When false, each join attempt
     * creates a new point-to-point session.
     */
    public boolean isMultipoint;

    /**
     * Holds the proximity for this SessionOpt
     */
    public byte proximity;

    /**
     * Holds the allowed transports for this SessionOpt
     */
    public short transports;

    /**
     * Construct a new session options class using the most commonly used options:
     * traffic = TRAFFIC_MESSAGES, isMultipoint = false, proximity = PROXIMITY_ANY,
     * transports = TRANSPORT_ANY.
     */
    public SessionOpts() {
        traffic = TRAFFIC_MESSAGES;
        isMultipoint = false;
        proximity = PROXIMITY_ANY;
        transports = TRANSPORT_ANY;
    }

    /**
     * Construct a new session options class using the provided options.
     *
     * @param traffic the traffic bitfield values.
     * @param isMultipoint the isMultipoint option
     * @param proximity the proximity bitfield values.
     * @param transports the transports bitfield values.
     */
    public SessionOpts(byte traffic, boolean isMultipoint, byte proximity, short transports) {
        this.traffic = traffic;
        this.isMultipoint = isMultipoint;
        this.proximity = proximity;
        this.transports = transports;
    }

    /**
     * Use reliable message-based communication to move data between session endpoints.
     */
    public static final byte TRAFFIC_MESSAGES       = 0x01;

    /**
     * Use unreliable (e.g., UDP) socket-based communication to move data between
     * session endpoints.  RAW does not imply raw sockets (that bypass ALL
     * enapsulation possibly down to the MAC level), it implies raw in an AllJoyn
     * sense --MESSAGE encapsulation is not used, but for example UDP + IP + MAC
     * encapsulation is used.
     */
    public static final byte TRAFFIC_RAW_UNRELIABLE = 0x02;

    /**
     * Use reliable (e.g., TCP) socket-based communication to move data between
     * session endpoints.  RAW does not imply raw sockets (that bypass ALL
     * enapsulation possibly down to the MAC level), it implies raw in an AllJoyn
     * sense --MESSAGE encapsulation is not used, but for example UDP + IP + MAC
     * encapsulation is used.
     */
    public static final byte TRAFFIC_RAW_RELIABLE   = 0x04;

    /**
     * Do not limit the spatial scope of sessions.  This means that sessions may
     * be joined by jointers located anywhere.
     */
    public static final byte PROXIMITY_ANY      = (byte)0xff;

    /**
     * Limit the spatial scope of sessions to the local host.  Interpret as
     * "the same physical machine."  This means that sessions may be joined by
     * jointers located only on the same physical machine as the one hosting the
     * session.
     */
    public static final byte PROXIMITY_PHYSICAL = 0x01;

    /**
     * Limit the spatial scope of sessions to anwhere on the local logical
     * network segment.  This means that sessions may be joined by jointers
     * located somewhere on the network.
     */
    public static final byte PROXIMITY_NETWORK  = 0x02;

    /**
     * Use no transport to communicate with a given session.
     */
    public static final short TRANSPORT_NONE      = 0x0000;

    /**
     * Use only the local transport to communicate with a given session.
     */
    public static final short TRANSPORT_LOCAL     = 0x0001;

    /**
     * Use only a wireless local area network to communicate with a given session.
     */
    public static final short TRANSPORT_WLAN      = 0x0004;

    /**
     * Transport using TCP (same as TRANSPORT_WLAN)
     */
    public static final short TRANSPORT_TCP       = TRANSPORT_WLAN;

    /**
     * Use only a wireless wide area network to communicate with a given session.
     */
    public static final short TRANSPORT_WWAN      = 0x0008;

    /**
     * Use only a wired local area network to communicate with a given session.
     */
    public static final short TRANSPORT_LAN      = 0x0010;

    /**
     * Use only the Wi-Fi Direct transport to communicate with a given session.
     */
    public static final short TRANSPORT_WFD      = 0x0080;

    /**
     * Use only the AllJoyn Reliable Datagram Protocol (flavor of reliable UDP)
     * to communicate with a given session.
     */
    public static final short TRANSPORT_UDP      = 0x0100;

    /**
     * Use any available transport to communicate with a given session.
     *
     * It is the case that (1) certain topologies of AllJoyn distributed
     * applications can cause problems when run on Wi-Fi Direct substrate
     * networks; (2) the specifics of authentication in Wi-Fi Direct networks
     * can also produce surprising results in some AllJoyn topologies; and (3)
     * there are implementation problems in existing Wi-Fi Direct systems that
     * prevent certain AllJoyn topologies from forming.  Because these issues
     * might produce surprising results in existing applications that are
     * unaware of the limitations, we do no enable Wi-Fi Direct automatically.
     *
     * Selecting ANY transport really means selecting ANY but Wi-Fi Direct.  In
     * order to enable discovery over Wi-Fi Direct, and use of the Wi-Fi Direct
     * transport an application author must positively act and OR in the
     * TRANSPORT_WFD bit.
     */
    public static final short TRANSPORT_ANY       = (short)0xffff & ~TRANSPORT_WFD;

    /**
     * Use any available IP-based transport to communicate with a given session.
     *
     * Selecting the IP transport really implies letting the system decice which
     * transport is best.
     */
    public static final short TRANSPORT_IP       = (TRANSPORT_TCP | TRANSPORT_UDP);

    public String toString( ) {
        StringBuilder result = new StringBuilder();
        result.append(this.getClass().getName() + " {");

        result.append("traffic = ");
        String value = String.format("(0x%02x)", traffic);
        result.append(value);
        if ((traffic & TRAFFIC_MESSAGES) != 0) result.append(" TRAFFIC_MESSAGES");
        if ((traffic & TRAFFIC_RAW_UNRELIABLE) != 0) result.append(" TRAFFIC_RAW_UNRELIABLE");
        if ((traffic & TRAFFIC_RAW_RELIABLE) != 0) result.append(" TRAFFIC_RAW_RELIABLE");

        result.append(", isMultipoint = ");
        value = String.format("%b", isMultipoint);
        result.append(value);

        result.append(", proximity =");
        value = String.format("(0x%02x)", proximity);
        result.append(value);
        if ((proximity & PROXIMITY_PHYSICAL) != 0) result.append(" PROXIMITY_PHYSICAL");
        if ((proximity & PROXIMITY_NETWORK) != 0) result.append(" PROXIMITY_NETWORK");

        result.append(", transports =");
        value = String.format("(0x%04x)", transports);
        result.append(value);
        if ((transports & TRANSPORT_LOCAL) != 0) result.append(" TRANSPORT_LOCAL");
        if ((transports & TRANSPORT_WLAN) != 0) result.append(" TRANSPORT_WLAN");
        if ((transports & TRANSPORT_WWAN) != 0) result.append(" TRANSPORT_WWAN");
        if ((transports & TRANSPORT_WFD) != 0) result.append(" TRANSPORT_WFD");

        result.append("}");
        return result.toString();
    }
}
